pkg/kube/wait.go

```diff
@@ -125,155 +107,175 @@ func (w *waiter) waitForResources(created Result) error {
-				if err != nil {
-					return false, err
-				}
-				pods = append(pods, list...)
-			case *v1.PersistentVolumeClaim:
-				claim, err := w.c.CoreV1().PersistentVolumeClaims(value.Namespace).Get(value.Name, metav1.GetOptions{})
-				if err != nil {
-					return false, err
-				}
-				pvc = append(pvc, *claim)
-			case *v1.Service:
-				svc, err := w.c.CoreV1().Services(value.Namespace).Get(value.Name, metav1.GetOptions{})
-				if err != nil {
-					return false, err
-				}
-				services = append(services, *svc)
+				ok, err = w.podsReadyForObject(value.Namespace, value)
+			}
+			if !ok || err != nil {
+				return false, err
 			}
 		}
-		isReady := w.podsReady(pods) && w.servicesReady(services) && w.volumesReady(pvc) && w.deploymentsReady(deployments)
-		return isReady, nil
+		return true, nil
 	})
 }
 
-func (w *waiter) podsReady(pods []v1.Pod) bool {
+func (w *waiter) podsReadyForObject(namespace string, obj runtime.Object) (bool, error) {
+	pods, err := w.podsforObject(namespace, obj)
+	if err != nil {
+		return false, err
+	}
 	for _, pod := range pods {
-		if !IsPodReady(&pod) {
-			w.log("Pod is not ready: %s/%s", pod.GetNamespace(), pod.GetName())
-			return false
+		if !w.isPodReady(&pod) {
+			return false, nil
 		}
 	}
-	return true
+	return true, nil
+}
+
+func (w *waiter) podsforObject(namespace string, obj runtime.Object) ([]corev1.Pod, error) {
+	selector, err := selectorsForObject(obj)
+	if err != nil {
+		return nil, err
+	}
+	list, err := getPods(w.c, namespace, selector.String())
+	return list, err
 }
 
-// IsPodReady returns true if a pod is ready; false otherwise.
-func IsPodReady(pod *v1.Pod) bool {
+// isPodReady returns true if a pod is ready; false otherwise.
+func (w *waiter) isPodReady(pod *corev1.Pod) bool {
 	for _, c := range pod.Status.Conditions {
-		if c.Type == v1.PodReady && c.Status == v1.ConditionTrue {
+		if c.Type == corev1.PodReady && c.Status == corev1.ConditionTrue {
 			return true
 		}
 	}
+	w.log("Pod is not ready: %s/%s", pod.GetNamespace(), pod.GetName())
 	return false
 }
 
-func (w *waiter) servicesReady(svc []v1.Service) bool {
-	for _, s := range svc {
-		// ExternalName Services are external to cluster so helm shouldn't be checking to see if they're 'ready' (i.e. have an IP Set)
-		if s.Spec.Type == v1.ServiceTypeExternalName {
-			continue
-		}
-
-		// Make sure the service is not explicitly set to "None" before checking the IP
-		if s.Spec.ClusterIP != v1.ClusterIPNone && !IsServiceIPSet(&s) {
-			w.log("Service is not ready: %s/%s", s.GetNamespace(), s.GetName())
-			return false
-		}
+func (w *waiter) serviceReady(s *corev1.Service) bool {
+	// ExternalName Services are external to cluster so helm shouldn't be checking to see if they're 'ready' (i.e. have an IP Set)
+	if s.Spec.Type == corev1.ServiceTypeExternalName {
+		return true
+	}
+	// Make sure the service is not explicitly set to "None" before checking the IP
+	if s.Spec.ClusterIP != corev1.ClusterIPNone && !isServiceIPSet(s) ||
```

Comment:
should this be an `&&` at the end? From what I can read, the old logic was "if cluster IP is NOT "none" and service IP is not set, AND if the service has a load balancer with an ingress defined: the service is not ready".

Reply:
may need to look into that.  All this changed was `s/v1/corev1/`

Reply:
Because reading diffs on github can be hard:

old code: https://github.com/helm/helm/blob/dev-v3/pkg/kube/wait.go#L233
new code: https://github.com/adamreese/helm/blob/v3/kube-interface/pkg/kube/wait.go#L196-L202

it looks like I was wrong; I thought the second `if` code was nested within the first.

Looks good here!